snuba/query/mql/parser.py

```diff
@@ -0,0 +1,636 @@
+        return str(node.text[1:-1])
+
+    def visit_string_tuple(self, node: Node, children: Sequence[Any]) -> Sequence[str]:
+        _, _, first, zero_or_more_others, _, _ = children
+        return [first[0], *(v[0] for _, _, _, v in zero_or_more_others)]
+
+    def visit_group_by_name(self, node: Node, children: Sequence[Any]) -> str:
+        return node.text
+
+    def visit_group_by_name_tuple(
+        self, node: Node, children: Sequence[Any]
+    ) -> Sequence[str]:
+        _, _, first, zero_or_more_others, _, _ = children
+        return [first, *(v for _, _, _, v in zero_or_more_others)]
+
+    def visit_target(self, node: Node, children: Sequence[Any]) -> Mapping[str, Any]:
+        target = children[0]
+        if isinstance(children[0], list):
+            target = children[0][0]
+        return target
+
+    def visit_variable(self, node: Node, children: Sequence[Any]) -> Any:
+        raise InvalidQueryException("Variables are not supported yet")
+
+    def visit_nested_expression(
+        self, node: Node, children: Sequence[Any]
+    ) -> Mapping[str, Any]:
+        return children[2]
+
+    def visit_aggregate(
+        self, node: Node, children: Sequence[Any]
+    ) -> SelectedExpression:
+        aggregate_name, zero_or_one = children
+        _, _, target, zero_or_more_others, *_ = zero_or_one
+
+        if "mri" in target:
+            metric_name = target["mri"]
+        elif "public_name" in target:
+            metric_name = target["public_name"]
+        else:
+            metric_name = ""
+        target["aggregate"] = SelectedExpression(
+            name=f"{aggregate_name}({metric_name})",
+            expression=FunctionCall(
+                alias=None,
+                function_name=aggregate_name,
+                parameters=[Column(alias=None, table_name=None, column_name="value")],
+            ),
+        )
+        return target
+
+    def visit_aggregate_name(self, node: Node, children: Sequence[Any]) -> str:
+        return node.text
+
+    def visit_quoted_mri(
+        self, node: Node, children: Sequence[Any]
+    ) -> Mapping[str, str]:
+        return {"mri": str(node.text[1:-1])}
+
+    def visit_unquoted_mri(
+        self, node: Node, children: Sequence[Any]
+    ) -> Mapping[str, str]:
+        return {"mri": str(node.text)}
+
+    def visit_quoted_public_name(
+        self, node: Node, children: Sequence[Any]
+    ) -> Mapping[str, str]:
+        return {"public_name": str(node.text[1:-1])}
+
+    def visit_unquoted_public_name(
+        self, node: Node, children: Sequence[Any]
+    ) -> Mapping[str, str]:
+        return {"public_name": str(node.text)}
+
+    def visit_identifier(self, node: Node, children: Sequence[Any]) -> str:
+        return node.text
+
+    def generic_visit(self, node: Node, children: Sequence[Any]) -> Any:
+        """The generic visit method."""
+        return children
+
+
+def parse_mql_query_initial(
+    body: str,
+    mql_context: Mapping[str, Any],
```

Comment:
Good point. Right now, there is visitor which validates and serializes a `MQLContext` object into a json on the sentry side. We should do the reverse and validate again on the snuba side. I can update this in this PR once we release 2.0.11 of snuba-sdk https://github.com/getsentry/snuba-sdk/blob/main/CHANGES.rst#2011 since it depends on that.

Reply:
@evanh, @volokluev I replicated the `MQLContext` class and its validator in snuba. However, I realized that it requires a bunch of other classes like `Rollup`, `MetricScope`, `Limit`, and `Offset` to also all be replicated in snuba. While this keeps snuba independency from snuba-sdk, there is a lot of duplicated code which will require maintenance if ever we change these classes. What do you think?

Reply:
I think we should duplicate the structures. I would rather have duplicates than tie Snuba to the SDK. Snuba has to duplicate a lot of SDK code anyways by necessity, so this has some precedence.